ext/json/json_encoder.c

```diff
@@ -29,6 +29,260 @@
 #include "zend_enum.h"
 #include "zend_property_hooks.h"
 #include "zend_lazy_objects.h"
+#include "zend_bitset.h"
+
+#if defined(ZEND_INTRIN_SSE4_2_NATIVE) || defined(ZEND_INTRIN_SSE4_2_FUNC_PROTO)
+# include <nmmintrin.h>
+#endif
+#ifdef ZEND_INTRIN_SSE4_2_FUNC_PROTO
+# include "zend_cpuinfo.h"
+#endif
+
+#ifdef __SSE2__
+# define JSON_USE_SIMD
+#endif
+
+typedef enum php_json_simd_result {
+	PHP_JSON_STOP,
+	PHP_JSON_SLOW,
+	PHP_JSON_NON_ASCII,
+} php_json_simd_result;
+
+/* Specialization of smart_str_appendl() to avoid performance loss due to code bloat */
+static zend_always_inline void php_json_append(smart_str *dest, const char *src, size_t len)
+{
+	/* dest has a minimum size of the input length,
+	 * this avoids generating initial allocation code */
+	ZEND_ASSERT(dest->s);
```

Comment:
I don't exactly how assert can avoid generating initial allocation if it's removed in production code..?

Reply:
> if it's removed in production code..?

Because it's not removed: it's changed to a ZEND_ASSUME on compilers that support this, and it's removed on compilers that don't. So it can be used as an optimization hint.

Reply:
Ah ok, it might make sense to integrate it to smart_str if it has some possitive impact as the same case applies on serialization and possibly other places.